File: templates/cpp_classes.cpp
#include "cpp_classes.hpp"
///
//! for include in includes:
#include "/*{include}*/"
//! endfor
///
//! for c_name, methods in c_methods:
//! for m_name, m_name_1, m_name_2, m_args in methods:
extern PyObject * /*{m_name_1}*/(/*{c_name}*/ * self, PyObject * const * args, Py_ssize_t nargs);
//! endfor
//! if not loop.last
///
//! endif
//! endfor
///
#if PY_VERSION_HEX >= 0x03070000
///
//! for c_name, methods in c_methods:
PyMethodDef /*{c_name}*/_methods[] = {
    //! for m_name, m_name_1, m_name_2, m_args in methods:
    {"/*{m_name}*/", (PyCFunction)/*{m_name_1}*/, /*{'METH_FASTCALL' if m_args else 'METH_NOARGS'}*/, 0},
    //! endfor
    {0},
};
//! if not loop.last
///
//! endif
//! endfor
///
#else
///
//! for c_name, methods in c_methods:
//! for m_name, m_name_1, m_name_2, m_args in methods:
//! if m_args
PyObject * /*{m_name_2}*/(/*{c_name}*/ * self, PyObejct * args) {
    return /*{m_name_1}*/(self, ((PyTupleObject *)args)->ob_item, ((PyVarObject *)args)->ob_size);
}
//! if not loop.last
///
//! endif
//! endif
//! endfor
//! if not loop.last
///
//! endif
//! endfor
///
//! for c_name, methods in c_methods:
PyMethodDef /*{c_name}*/_methods[] = {
    //! for m_name, m_name_1, m_name_2, m_args in methods:
    {"/*{m_name}*/", (PyCFunction)/*{m_name_2}*/, /*{'METH_VARARGS' if m_args else 'METH_NOARGS'}*/, 0},
    //! endfor
    {0},
};
//! if not loop.last
///
//! endif
//! endfor
///
#endif
///
//! for c_name in c_names:
PyType_Slot /*{c_name}*/_slots[] = {
    {Py_tp_methods, /*{c_name}*/_methods},
    {0},
};
//! if not loop.last
///
//! endif
//! endfor
///
//! for c_name in c_names:
PyType_Spec /*{c_name}*/_spec = {
    mgl_name "./*{c_name[3:]}*/",
    sizeof(/*{c_name}*/),
    0,
    Py_TPFLAGS_DEFAULT,
    /*{c_name}*/_slots,
};
//! if not loop.last
///
//! endif
//! endfor
///

File: templates/cpp_classes.py
import glob
import os
import re

import jinja2

if not os.path.isdir('moderngl/mgl'):
    raise FileNotFoundError('run this from the project root')

re_methods = []
includes = []

for fn in glob.glob('moderngl/mgl/**/*.cpp', recursive=True):
    with open(fn) as f:
        code = f.read()

    re_methods += re.findall(r'^PyObject \* ([A-Za-z0-9]+)_meth_([A-Za-z0-9]+)\(\1 \* self(, PyObject \* const \* args, Py_ssize_t nargs)\) \{$', code, re.M)

for fn in glob.glob('moderngl/mgl/*.hpp'):
    with open(fn) as f:
        code = f.read()

    if 'struct MGL' in code:
        includes.append(os.path.relpath(fn, 'moderngl/mgl/generated').replace('\\', '/'))

c_names = sorted(set(c_name for c_name, m_name, args in re_methods))
c_methods = []

for name in c_names:
    methods = []
    for c_name, m_name, args in re_methods:
        if c_name == name:
            m_name_1 = f'{c_name}_meth_{m_name}'
            m_name_2 = f'{m_name_1}_va' if args else m_name_1
            methods.append((m_name, m_name_1, m_name_2, bool(args)))
    c_methods.append((name, methods))

env = jinja2.Environment(
    block_start_string = '/*[',
    block_end_string = ']*/',
    variable_start_string = '/*{',
    variable_end_string = '}*/',
    comment_start_string = '/*#',
    comment_end_string = '#*/',
    line_statement_prefix = '//!',
    line_comment_prefix = '///',
)

with open('templates/cpp_classes.cpp') as f:
    template = f.read()

with open('moderngl/mgl/generated/cpp_classes.cpp', 'w') as f:
    f.write(env.from_string(template).render(c_methods=c_methods, c_names=c_names, includes=includes))

with open('templates/cpp_classes.hpp') as f:
    template = f.read()

with open('moderngl/mgl/generated/cpp_classes.hpp', 'w') as f:
    f.write(env.from_string(template).render(c_methods=c_methods, c_names=c_names))
